### DISCUSSION TITLE
"Datasource unauthorized" for private Nuget feed

### DISCUSSION THREAD

**Discussion by ptr1120** [renovatebot/renovate#8576] [Request Help] [answered]
**What Renovate type, platform and version are you using?**

I am using self-hosted Renovate together with GitLab and gitlab-runner configuration from [here](https://gitlab.com/renovate-bot/renovate-runner). 
It is a docker container of renovate used with version renovate/renovate:24.35.2.
In the renovate.json I have:
```
...
 "nuget": {
    "registryUrls": [
      "https://api.nuget.org/v3/index.json",
      "https://mydomain/nuget/mycompany/v3/index.json"
    ]
  }
```

I am starting the runner with:
```
renovate --dry-run=true --host-rules '{"hostType":"nuget","domainName":"mydomain","username":"$PRIVATE_NUGET_USERNAME","password":"$PRIVATE_NUGET_PASSWORD"}'
```

**Describe the bug**

During running Renovate updates all Nugets from public Nuget feed correctly. But Nugets from private feed do not get updated. The debug logs show "Datasource unauthorized" for Nugets from private feed.
If I try to access the private Nuget with the same Url as in the log and with basicAuth together with curl I get a successful response.
```
curl -u username:password --request GET https://mydomain/nuget/mycompany/v3/registrations-gz/mynugetpackage/index.json
```

**Relevant debug logs**
```
{"name":"renovate","hostname":"runner-rbt1jkos-project-200-concurrent-0","pid":14,"level":20,"logContext":"MW-cFyoD6","config":{"repositories":["myrepository"],"dryRun":true,"hostRules":["{\"hostType\":\"nuget\"","\"domainName\":\"mydomain\"","\"username\":\"username\"","\"password\":\"password\"}"]},"msg":"CLI config","time":"2021-02-07T17:16:34.724Z","v":0}
..
{"name":"renovate","hostname":"runner-rbt1jkos-project-200-concurrent-0","pid":14,"level":20,"logContext":"MW-cFyoD6","repository":"myrepository","datasource":"nuget","lookupName":"mynugetpackage","url":"https://mydomain/nuget/mycompany/v3/registrations-gz/mynugetpackage/index.json","msg":"Datasource unauthorized","time":"2021-02-07T17:17:48.106Z","v":0}
{"name":"renovate","hostname":"runner-rbt1jkos-project-200-concurrent-0","pid":14,"level":20,"logContext":"MW-cFyoD6","repository":"myrepository","dependency":"mynugetpackage","packageFile":"myproject.csproj","msg":"Failed to look up dependency mynugetpackage","time":"2021-02-07T17:17:48.106Z","v":0}
```

The username and password are correct in the logs, the same as I tried with curl.

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/master/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] This is a really small bug, it does not need a reproduction (think small typo)
- [ ] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [x] I cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

Any idea or help would be great, thank you.

**Comment by rarkins** [comment, marked as answer]
If you're seeing text like `["{\"h` with the escaped quote then it means that the hostRules have not been parsed as JSON and instead as strings. Try wrapping in `[]` because `hostRules` should be an array, not an object:

```
renovate --dry-run=true --host-rules '[{"hostType":"nuget","domainName":"mydomain","username":"$PRIVATE_NUGET_USERNAME","password":"$PRIVATE_NUGET_PASSWORD"}]'
```

Unfortunately CLI parsing can be challenging - a `config.js` could be easier if that doesn't work

**Comment by ptr1120** [comment]
Thank you very much, that solved my failure.

------ Originalnachricht ------
Von: "Rhys Arkins" <notifications@github.com>
An: "renovatebot/renovate" <renovate@noreply.github.com>
Cc: "Peter Bruch" <peta1120@googlemail.com>; "Author" 
<author@noreply.github.com>
Gesendet: 07.02.2021 19:06:55
Betreff: Re: [renovatebot/renovate] "Datasource unauthorized" for 
private Nuget feed (#8576)
